File: dice.lua
#!/usr/bin/env lua


function inspect (t)
    for i, xi in pairs(t) do
	print(i, xi) 
    end
end


function main (argv)
    local rolls, constant = parse(argv)

    local min = 0
    local max = 0
    local combinations = 1;
    for d,n in pairs(rolls) do
	min = min + n
	max = max + n * d
	combinations = combinations * d^n
    end

    local polys = mkpolys(rolls)

    local biggun = { 1 }
 
    for i, x in pairs(polys) do
	biggun = polymul(biggun, x)
    end

    for x = min, max+1 do
	local coeff = biggun[x]

	if coeff then
	    local px    = coeff / combinations
	    local hg    = histogram(px)
	    print( string.format("%d\t\t%.5f\t\t%s", x-1+constant, px, hg) )
	end
    end
end


function parse (args)
    local pat   = "(%d+)d(%d+)"
    local rv    = {}
    local adder = 0;

    for i, xi in ipairs(arg) do
	if (string.find(xi, pat)) then
	    _, _, n, d = string.find(xi, pat)
	    rv[d] = rv[d] or 0
	    rv[d] = rv[d] + n
	else
	    adder = adder + xi
	end
    end

    return rv, adder
end


function mkpolys (rolls)
    local polys = {}

    for d, n in pairs(rolls) do
	for i = 1, n do
	    local p = {}

	    for j = 1, d do
		p[j] = 1
	    end

	    table.insert(polys, p)
	end
    end

    return polys
end


function polymul (x, y)
    local rv = {};

    for i, xi in pairs(x) do
	for j, yj in pairs(y) do
	    local cell = i + j
	    if rv[cell] == nil then
		rv[cell] = 0;
	    end

	    rv[cell] = rv[cell] + (xi * yj)
	end
    end

    return rv
end


function histogram (n)
    local num = (0.5 + (500 * n));

    local tmp = {}
    for i = 0, num do
	tmp[i] = "#"
    end

    return table.concat(tmp, "")
end


main(arg)

File: dice.rb
#!/usr/bin/env ruby


class Roll
    def initialize (n, d)
	@n = n
	@d = d
    end

    attr_reader :n
    attr_reader :d
end


class Polynomial
    def initialize (yarr)
	@coefficients = yarr
    end

    attr_reader :coefficients

    def self.mkroll (n)
	x = [0]

	(1..n).each do |i|
	    x[i] = 1
	end

	Polynomial.new(x)
    end

    def *(other)
	noob = []

	@coefficients.each_with_index do |xi, i|
	    other.coefficients.each_with_index do |yj, j|
		cell = i + j

		if noob[cell].nil?
		    noob[cell] = 0
		end

		noob[cell] += xi * yj
	    end
	end

	Polynomial.new(noob)
    end
end


def parse (args)
    rolls = []
    adder = 0

    args.each do |arg|
	if arg.match("d")
	    rolls.push(
		Roll.new(
		    *arg.split("d").map {|x| x.to_i }
		)
	    )
	else
	    adder += arg.to_i
	end
    end

    return [rolls, adder]
end


def histogram (n)
    num  = (0.5 + (500 * n));
    "#" * num
end


def main (args)
    rolls, constant = parse(args)

    min = 0
    max = 0
    combinations = 1;

    polys = []

    rolls.each do |roll|
	min += roll.n
	max += roll.n * roll.d
	combinations *= roll.d**roll.n

	(1..roll.n).each do |r|
	    polys.push( Polynomial.mkroll(roll.d) )
	end
    end

    biggun = Polynomial.new([1]) # identity

    polys.each do |p|
	biggun *= p
    end

    (min..max).each do |deg|
	coeff = biggun.coefficients[deg]

	if !(coeff.nil? or coeff == 0)
	    px    = coeff.to_f / combinations.to_f
	    hg    = histogram(px)
	    printf("%d\t\t%.5f\t\t%s\n", deg + constant, px, hg)
	end
    end
end


main(ARGV)


